# Tutorial - Text Mining - Classification 

We will predict the category of discussion posts in a newsgroup.

**The unit of analysis is a discussion post**

### Import common packages

In [56]:
import pandas as pd
import numpy as np
import nltk
nltk.download('averaged_perceptron_tagger') # you only need to run this once
nltk.download('punkt')
nltk.download('wordnet')
nltk.download('omw-1.4')
np.random_seed = 1

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\suman\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\suman\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\suman\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\suman\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


### Load data

In [57]:
news = pd.read_csv('news.csv')

news.shape


(597, 5)

In [58]:
news.head(5)

TEXT  graphics  hockey  \
0  I have a few reprints left of chapters from my...         1       0   
1  gnuplot, etc. make it easy to plot real valued...         1       0   
2  Article-I.D.: snoopy.1pqlhnINN8k1 References: ...         1       0   
3  Hello, I am looking to add voice input capabil...         1       0   
4  I recently got a file describing a library of ...         1       0   

   medical newsgroup  
0        0  graphics  
1        0  graphics  
2        0  graphics  
3        0  graphics  
4        0  graphics

### Check for missing values

In [59]:
news[['TEXT']].isna().sum()

TEXT    0
dtype: int64

## Assign the input variable to X and the target variable to y

In [60]:
X = news['TEXT']
X

0      I have a few reprints left of chapters from my...
1      gnuplot, etc. make it easy to plot real valued...
2      Article-I.D.: snoopy.1pqlhnINN8k1 References: ...
3      Hello, I am looking to add voice input capabil...
4      I recently got a file describing a library of ...
                             ...                        
592    carl@SOL1.GPS.CALTECH.EDU (Carl J Lydick) writ...
593    In article < 1qmlgaINNjab@hp-col.col.hp.com> ,...
594    Article-I.D.: kestrel.1993Apr16.172052.27843 R...
595    In article < 1qmlgaINNjab@hp-col.col.hp.com> ,...
596    I have a 42 yr old male friend, misdiagnosed a...
Name: TEXT, Length: 597, dtype: object

This is a multi-class classification problem. There are three categories we will predict:<br>
Whether a post is "graphics," "hockey," or "medical" related

In [61]:
y = news['newsgroup']
y.unique()

array(['graphics', 'hockey', 'medical'], dtype=object)

In [62]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()
le.fit(y)
print(le.classes_)
y = le.transform(y)

y


['graphics' 'hockey' 'medical']


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [63]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Like CountVectorizer, TfidfVectorizer will covert to lowercase, remove punctuation, and remove 
# stop words - to remove other things, such as numbers, use the token_pattern parameter
vectorizer = TfidfVectorizer(stop_words='english', lowercase=True, token_pattern="[^\W\d_]+")

Xt = vectorizer.fit_transform(X)

df = pd.DataFrame(Xt.toarray(), columns=vectorizer.get_feature_names_out())
df

aa  aalborg  aamrl  aangeboden  aantal   aaplay  aarnet        ab  abad  \
0    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
1    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
2    0.0      0.0    0.0         0.0     0.0  0.09173     0.0  0.000000   0.0   
3    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
4    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.126841   0.0   
..   ...      ...    ...         ...     ...      ...     ...       ...   ...   
592  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
593  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
594  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
595  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   
596  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.000000   0.0   

     abandon  ...  zoo  zool  zorn   zt   zu  zubov  zupancic  zurich  zyeh  \
0        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
1        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
2        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
3        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
4        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
..       ...  ...  ...   ...   ...  ...  ...    ...       ...     ...   ...   
592      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
593      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
594      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
595      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
596      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   

     zzz  
0    0.0  
1    0.0  
2    0.0  
3    0.0  
4    0.0  
..   ...  
592  0.0  
593  0.0  
594  0.0  
595  0.0  
596  0.0  

[597 rows x 12289 columns]

## Lemmatization


In [64]:
import nltk
from nltk.stem import WordNetLemmatizer 
from nltk import pos_tag, word_tokenize

In [65]:
corpus = X
corpus

0      I have a few reprints left of chapters from my...
1      gnuplot, etc. make it easy to plot real valued...
2      Article-I.D.: snoopy.1pqlhnINN8k1 References: ...
3      Hello, I am looking to add voice input capabil...
4      I recently got a file describing a library of ...
                             ...                        
592    carl@SOL1.GPS.CALTECH.EDU (Carl J Lydick) writ...
593    In article < 1qmlgaINNjab@hp-col.col.hp.com> ,...
594    Article-I.D.: kestrel.1993Apr16.172052.27843 R...
595    In article < 1qmlgaINNjab@hp-col.col.hp.com> ,...
596    I have a 42 yr old male friend, misdiagnosed a...
Name: TEXT, Length: 597, dtype: object

In [66]:

new_corpus = []
wnl = WordNetLemmatizer()
for document in corpus:
    new_document = ""
    for word, tag in pos_tag(word_tokenize(document)):
        wntag = tag[0].lower()
        wntag = wntag if wntag in ['a', 'r', 'n', 'v'] else None
        if not wntag:
            lemma = word
        else:
            lemma = wnl.lemmatize(word, wntag)
        new_document+= lemma + " "
    new_corpus += [new_document]



In [67]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer(stop_words='english', lowercase=True, token_pattern="[^\W\d_]+")

Xs = vectorizer.fit_transform(new_corpus)

df = pd.DataFrame(Xs.toarray(), columns=vectorizer.get_feature_names_out())
df

aa  aalborg  aamrl  aangeboden  aantal   aaplay  aarnet      ab  abad  \
0    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
1    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
2    0.0      0.0    0.0         0.0     0.0  0.09234     0.0  0.0000   0.0   
3    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
4    0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.1337   0.0   
..   ...      ...    ...         ...     ...      ...     ...     ...   ...   
592  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
593  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
594  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
595  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   
596  0.0      0.0    0.0         0.0     0.0  0.00000     0.0  0.0000   0.0   

     abandon  ...  zoo  zool  zorn   zt   zu  zubov  zupancic  zurich  zyeh  \
0        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
1        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
2        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
3        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
4        0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
..       ...  ...  ...   ...   ...  ...  ...    ...       ...     ...   ...   
592      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
593      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
594      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
595      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   
596      0.0  ...  0.0   0.0   0.0  0.0  0.0    0.0       0.0     0.0   0.0   

     zzz  
0    0.0  
1    0.0  
2    0.0  
3    0.0  
4    0.0  
..   ...  
592  0.0  
593  0.0  
594  0.0  
595  0.0  
596  0.0  

[597 rows x 10925 columns]

### By using lemmatization, 597 rows × 12289 columns reduced to 597 rows × 10925 columns.

## Split the data

In [68]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(Xs, y, test_size=0.3)

In [69]:
X_train.shape, y_train.shape

((417, 10925), (417,))

In [70]:
X_test.shape, y_test.shape

((180, 10925), (180,))

In [71]:
X_train

<417x10925 sparse matrix of type '<class 'numpy.float64'>'
	with 27888 stored elements in Compressed Sparse Row format>

In [72]:
X_test

<180x10925 sparse matrix of type '<class 'numpy.float64'>'
	with 13465 stored elements in Compressed Sparse Row format>

In [73]:
y_train[:5]

array([2, 2, 2, 2, 2])

## Sklearn: Text preparation

For simplicity (and focus), we will not do any text cleaning or preprocessing. We will just use the raw text as input to the model. See the text mining fundamentals tutorial for more details on text cleaning and preprocessing.

**Notice in the previous step that we use `fit_transform` on TRAIN. When we transform the TEST data, we need to use `transform` only. This enables us to keep the number of columns (features) the same across the data sets. Otherwise, they WILL be different, and no model will work!**

In [74]:
X_train.shape, X_test.shape

((417, 10925), (180, 10925))

In [75]:
# These data sets are "sparse matrix". We can't see them unless we convert using toarray()
X_train

<417x10925 sparse matrix of type '<class 'numpy.float64'>'
	with 27888 stored elements in Compressed Sparse Row format>

In [76]:
# These data sets are "sparse matrix". We can't see them unless we convert using toarray()
X_train.toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

## Latent Semantic Analysis (Singular Value Decomposition)

In [77]:
from sklearn.decomposition import TruncatedSVD

svd = TruncatedSVD(n_components=300, n_iter=10) #n_components is the number of topics, which should be less than the number of features

X_train= svd.fit_transform(X_train)
X_test = svd.transform(X_test)


In [78]:
X_train.shape, X_test.shape

((417, 300), (180, 300))

## Random Forest

In [79]:
from sklearn.ensemble import RandomForestClassifier 

rnd_clf = RandomForestClassifier(n_estimators=100, max_leaf_nodes=16, n_jobs=-1) 
_ = rnd_clf.fit(X_train, y_train)

### Evaluating Model Performance

In [80]:
from sklearn.metrics import accuracy_score

In [81]:
#Train accuracy - Not a good measure of model performance as we are using the same data set to train and test
y_pred_train = rnd_clf.predict(X_train)
acc = accuracy_score(y_train, y_pred_train)
print(f"Train acc: {accuracy_score(y_train, y_pred_train):.4f}")

Train acc: 0.9880


In [82]:
#Test accuracy
y_pred_test = rnd_clf.predict(X_test)
acc = accuracy_score(y_test, y_pred_test)
print(f"Train acc: {accuracy_score(y_test, y_pred_test):.4f}")

Train acc: 0.8667


In [83]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred_test)

array([[52,  0,  7],
       [ 1, 49,  0],
       [11,  5, 55]], dtype=int64)

## Stochastic Gradient Descent Classifier

In [84]:
from sklearn.linear_model import SGDClassifier

sgd_clf = SGDClassifier(max_iter=100)
_ = sgd_clf.fit(X_train, y_train)

### Evaluating Model Performance

In [85]:
#Train accuracy
y_pred_train = sgd_clf.predict(X_train)
print(f"Train acc: {accuracy_score(y_train, y_pred_train):.4f}")

Train acc: 1.0000


In [86]:
#Test accuracy
y_pred_test = sgd_clf.predict(X_test)
print(f"test acc: {accuracy_score(y_test, y_pred_test):.4f}")

test acc: 0.9111


In [87]:
# Confusion Matrix
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred_test)

array([[59,  0,  0],
       [ 2, 47,  1],
       [10,  3, 58]], dtype=int64)